x/fswap/keeper/keeper.go

```diff
@@ -6,37 +6,316 @@ import (
+
+	iterator := swappedDataStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		swapped := types.Swapped{}
+		k.cdc.MustUnmarshal(iterator.Value(), &swapped)
+		if cb(swapped) {
+			break
+		}
+	}
+}
+
+func (k Keeper) getSwapped(ctx sdk.Context) (types.Swapped, error) {
+	toDenom, err := k.toDenom(ctx)
+	if err != nil {
+		return types.Swapped{}, err
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	key := swappedKey(toDenom)
+	bz := store.Get(key)
+	if bz == nil {
+		return types.Swapped{}, types.ErrSwappedNotFound
+	}
+
+	swapped := types.Swapped{}
+	if err := k.cdc.Unmarshal(bz, &swapped); err != nil {
+		return types.Swapped{}, err
+	}
+	return swapped, nil
+}
+
+func (k Keeper) setSwapped(ctx sdk.Context, swapped types.Swapped) error {
+	store := ctx.KVStore(k.storeKey)
+	key := swappedKey(swapped.ToCoinAmount.Denom)
+	bz, err := k.cdc.Marshal(&swapped)
+	if err != nil {
+		return err
+	}
+
+	store.Set(key, bz)
+	return nil
+}
+
+func (k Keeper) getAllSwapInits(ctx sdk.Context) []types.SwapInit {
+	swapInits := make([]types.SwapInit, 0)
+	k.iterateAllSwapInits(ctx, func(swapInit types.SwapInit) bool {
+		swapInits = append(swapInits, swapInit)
+		return false
+	})
+	return swapInits
+}
+
+func (k Keeper) iterateAllSwapInits(ctx sdk.Context, cb func(swapped types.SwapInit) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	swapInitDataStore := prefix.NewStore(store, swapInitPrefix)
+
+	iterator := swapInitDataStore.Iterator(nil, nil)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		swapInit := types.SwapInit{}
+		k.cdc.MustUnmarshal(iterator.Value(), &swapInit)
+		if cb(swapInit) {
+			break
+		}
+	}
+}
+
+func (k Keeper) getSwappableNewCoinAmount(ctx sdk.Context) (sdk.Coin, error) {
+	swapCap, err := k.swapCap(ctx)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	swapped, err := k.getSwapped(ctx)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	denom, err := k.toDenom(ctx)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	remainingAmount := swapCap.Sub(swapped.GetToCoinAmount().Amount)
+
+	return sdk.NewCoin(denom, remainingAmount), nil
+}
+
+func (k Keeper) fromDenom(ctx sdk.Context) (string, error) {
+	swapInit, err := k.swapInit(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return swapInit.FromDenom, nil
+}
+
+func (k Keeper) toDenom(ctx sdk.Context) (string, error) {
+	swapInit, err := k.swapInit(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return swapInit.GetToDenom(), nil
+}
+
+func (k Keeper) swapMultiple(ctx sdk.Context) (sdk.Int, error) {
+	swapInit, err := k.swapInit(ctx)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return swapInit.SwapMultiple, nil
+}
+
+func (k Keeper) swapCap(ctx sdk.Context) (sdk.Int, error) {
+	swapInit, err := k.swapInit(ctx)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return swapInit.AmountCapForToDenom, nil
+}
+
+func (k Keeper) swapInit(ctx sdk.Context) (types.SwapInit, error) {
```

Comment:
Now the request `MsgSwap` and `MsgSwapAll` have enough information in its context, we can omit this finder. Applying this, you won't need denom verifications.